README.md
<img src=".github/media/VASS.png" width="200" alt="VASS Logo"/>

## System  
The system consists of several functions and configs which turn the Virtual Arsenal into a place where you can trade items for money and vice versa. A money system is not provided as of now. Instead you have the opportunity to implement your own one, like Ravage.<br/>
To make editing this system easier I have also created a mod which will let you set the attributes of a trader directly in 3den without creating a dependency:<br/>
https://steamcommunity.com/sharedfiles/filedetails/?id=1760193128

## Setup  
### Init
Copy the files from the master branch to your mission directory where the mission.sqm is located. The function which initializes the system is called automatically before each mission start. 

### Add trader
With `TER_fnc_addShop` you can register objects and units as traders.

### Add shop inventory
The function `TER_fnc_addShopCargo` handles the list of items which a trader sells or buys.
  
## Functions
This are the currently available functions which can be used to influence the system:<br/>
### TER_fnc_addShop
```
/*
	Author: Terra

	Description:
		Add shop action to an object. Any previous actions of this type are
		removed. VASS is only activated when the arsenal is opened via this
		function.

	Parameter(s):
		0:	OBJECT - Object to which the action is attached
		Optional:
		1:	STRING - Title of the action
			Default: "Shop"
		2:	NUMBER - Priority of the action, higher values mean the entry will appear first in the scroll wheel menu
			Default: 1.5
		3:	STRING - Condition for the addAction to show
			Default: "alive _this && alive _target"
		4:	NUMBER - Radius of the addAction
			Default: 5

	Returns:
		NUMBER - ID of the action, also saved as "TER_VASS_actionID" on the object

	Example(s):
		[cursorObject] call TER_fnc_addShop; //-> 0
		[cursorObject, "BLUFOR Weapon Shop", 999, "playerSide == west", 15] call TER_fnc_addShop; //-> 1
*/
```
### TER_fnc_addShopCargo
```
/*
	Author: Terra

	Description:
		Change the inventory of a shop.

	Parameter(s):
		0:	OBJECT - The object whose shop inventory will be changed
		1:	ARRAY - List of items, prices and amounts to add
				Format: ["class0", price, amount, "class1", price, amount,..., "classN", price, amount]
					Class: STRING - Class name of the item
					Price: NUMBER - The cost of the item
					Amount: NUMBER or BOOL - How many items the trader has. True means unlimited, false removes it from the inventory.
		Optional:
		2:	NUMBER - Overwrite mode:
				0 - Don't overwrite, only add new things
				1 - (default) Overwrite soft, only adjust prices and add new things
				2 - Hard overwrite, the passed array becomes the new inventory
				3 - Overwrite old, don't add new entries, only modify old ones
				4 - Amount diff, add/substract amounts
		3:	BOOL - BOOL - Change inventory for all players. If not specified, the _object's "TER_VASS_shared" variable is used. If this isn't set either it defaults to true.

	Returns:
		ARRAY - New shop cargo array

	Example(s):
		[cursorObject, ["50Rnd_570x28_SMG_03",50,40,"SMG_03_black",1000,5] call TER_fnc_addShopCargo; //-> ["50Rnd_570x28_SMG_03",50,40,"SMG_03_black",1000,5]
		[cursorObject, ["10Rnd_338_Mag",1000,5], 0, true] call TER_fnc_addShopCargo; //-> ["50Rnd_570x28_SMG_03",50,40,"SMG_03_black",1000,5,"10Rnd_338_Mag",1000,5]
		[cursorObject, ["10Rnd_338_Mag",1000,5], 2, true] call TER_fnc_addShopCargo; //-> ["10Rnd_338_Mag",1000,5]
*/
```
### TER_fnc_getItemValues
```
/*
	Author: Terra

	Description:
		Get the price and amount of an item in a shop.

	Parameter(s):
		0:	OBJECT - Shop object
		1:	STRING - Classname of the item
		Optional:
		2:	NUMBER - Type of return:
				0: class
				1: price
				2: amount
				-1: array of [class, price, amount]
			Default: -1
		3:	ARRAY - Default return when item is not in the shop
			Default: [parameter 0, 0, -1]

	Returns:
		ARRAY or STRING or NUMBER - Depending on param 2, the requested value

	Example(s):
		[cursorObject, "SMG_03_black"] call TER_fnc_getItemValues; //-> ["SMG_03_black", 1000, 5]
		[cursorObject, "SMG_03_black", 1] call TER_fnc_getItemValues; //-> 1000
		[cursorObject, "bogus", nil, ["bogus", 1234, 56]] call TER_fnc_getItemValues; //-> ["bogus", 1234, 56]
*/
```
### TER_fnc_resetTimer
```
/*
	Author: Terra

	Description:
		Change the items of a shop after a certain time.

	Parameter(s):
		0:	OBJECT - Shop object
		1:	ARRAY or BOOL - List of items to readd/remove or true to reset the inventory to default
		Optional:
		2:	NUMBER - Time until reset. Negative values will use the "TER_VASS_refresh" variable from the object.
			Default: -1
		3:	BOOL - Passed items will become the only ones after reset. If items is true then it is automatically set to reset.
			Default: false

	Returns:
		BOOL - true when done

	Example(s):
		[cursorObject, ["10Rnd_338_Mag",1000,5,"50Rnd_570x28_SMG_03",50,40,"SMG_03_black",1000,5]] spawn TER_fnc_resetTimer; //-> Script handle
		[cursorObject, ["10Rnd_338_Mag",1000,5,"50Rnd_570x28_SMG_03",50,40,"SMG_03_black",1000,5], 59, true] spawn TER_fnc_resetTimer; //-> Script handle
*/
```
### TER_fnc_VASShandler
```
/*
	Author: Terra

	Description:
		This is meant as an "API" of sorts. The following modes are available:
		"getMoney":
			Description:
				Get the money of the given unit
			Parameter(s):
				0:	OBJECT - The unit whose money is checked
			Returns:
				NUMBER - Money of the unit
		"setMoney":
			Description:
				Set the money of the given unit
			Parameter(s):
				0:	OBJECT - The unit whose money is modified
				1:	NUMBER - Change (not final amount, that is up to this function!)
			Returns:
				NOTHING - No return expected


	Parameter(s):
		0:	STRING - Mode
		Optional:
		1:	ARRAY - Parameters passed to the different modes
			Default: []

	Returns:
		ANY - Whatever the mode returns

	Example(s):
		["getMoney", [player]] call TER_fnc_VASShandler; //-> 0
		["setMoney", [player, -5900]] call TER_fnc_VASShandler; //-> nil
*/
```
### TER_fnc_getCurrency
```
/*
	Author: M1keSK

	Description:
		Get currency of shop.

	Parameter(s):
		0:	OBJECT - Shop object

	Returns:
		STRING - return currency symbol for shop

	Example(s):
		[cursorObject] call TER_fnc_getCurrency; //-> $
*/
```
### TER_fnc_setCurrency
```
/*
	Author: M1keSK

	Description:
		Set currency of shop.

	Parameter(s):
		0:	OBJECT - Shop object
		1:	STRING - currency symbol

	Returns:
		nothing

	Example(s):
		[cursorObject, "€"] call TER_fnc_setCurrency;
		[cursorObject, "£"] call TER_fnc_setCurrency;
		[cursorObject, "Ұ"] call TER_fnc_setCurrency;
*/
```

## Screenshots
### Ingame
<img src=".github/media/screenshot1.jpg" height="500"><br>
As you can see the prices of the items are listed right next to it in a green color. The lists are also sortable by either name, price from low to high and high to low.<br>
<br>
<img src=".github/media/screenshot2.jpg" height="500"><br>
When setting a maximum amount of items which can be added to one of the containers (uniform, vest, backpack) this maximum amount is also displayed next to the current amount. Items which are not part of the shop are set to a cost of 0.<br>
<br>
<img src=".github/media/screenshot3.jpg" height="500"><br>
Before leaving the shop the player can see what they will buy or sell. When leaving the shop any other way than over this menu then the loadout will be reset to the one with which he entered the shop.

### 3den Editor
<img src=".github/media/3denAttributes.jpg" height="500"><br>
The 3den mod adds a new section to the attributes in 3den. You can set
- Title
- Condition
- Radius
- Priority

of the addAction. Additionaly you can copy and paste the shop's inventory or edit it in the UI by clicking the "EDIT VASS INVENTORY" button. You can also set the time the shop takes to refresh (see [TER_fnc_resetTimer](#TER_fnc_resetTimer)) and the [locality](https://community.bistudio.com/wiki/Multiplayer_Scripting) of the whole shop.

<img src=".github/media/3denCargo.jpg" height="500"><br>
You can use the VASS Inventory UI to edit the shop's cargo array. You can include or exclude item types in your search by toggling them in the toolbox at the top. To apply your changes you have to confirm by clicking the OK button on the UI and then the OK button on the Attributes window in 3den.

# License
This project is licensed under the [GNU General Public License v3.0](https://choosealicense.com/licenses/gpl-3.0/).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.1.1] - 2022-08-02
### Changed
### Fixed
- Remote players could not see VASS action

## [2.1.0] - 2021-10-06
### Added
- Repository changelog
- License
- New UI
  - Seperated from the attribute's display
  - Filter functionality to include or exclude items
  - Search functionality
  - Attributes now include an edit box with the array to copy and paste from/to
- Stringtable
### Changed
- New project structure
  - Dedicated mission folder
  - Dedicated mod folder
  - hemtt compatible
  - Stopped using branches for different parts like a cavemen
- Private variables for performance
- Better function headers

## [1.0.0] - 2017-06-20


VASS_3den/Cfg3den.hpp
#include "\a3\3DEN\UI\macros.inc"
#include "gui\scripts\idcs.inc"
class ctrlButton;
class ctrlListNBox;
class ctrlStatic;
class ctrlToolbox;
class ctrlEdit;
class ctrlCheckboxBaseline;
class ctrlCombo;
class ctrlControlsGroup;
class ctrlStructuredText;
class Cfg3den
{
	/*
	class Mission
	{
		class Scenario
		{
			class AttributeCategories
			{
				class VASS_moneySystem
				{
					displayName = "VASS - Money System";
					class Attributes
					{
						class VASS_mission_money
						{
							displayName = "Enable VASS Money System";
							tooltip = "Enable or disable the use of the money system.";
							property = "VASS_moneyEnable";
							control = "Checkbox";
							expression = "\
								if (_value && !is3den) then {\
									if (isnil 'vass_db') then {vass_db = []};\
									[vass_db, ['settings', 'money', 'enable'], true] call BIS_fnc_dbValueSet;\
								};\
							";
							defaultValue = "false";
						};
					};
				};
			};
		};
	};
	*/
	class Object
	{
		class AttributeCategories
		{
			class VASS
			{
				displayName = "VASS";
				collapsed = 1;
				class Attributes
				{
					class addAction
					{
						displayName = "";
						tooltip = "";
						property = "VASS_addAction";
						control = "VASS_addAction";
						expression = "\
							_value = parseSimpleArray _value;\
							if (_value#0 && !is3DEN) then {[_this,_value#1,parseNumber (_value#4),_value#2,parseNumber (_value#3)] call TER_fnc_addShop};\
						";
						defaultValue = "str[false,""Shop"",""alive _this && alive _target"",""5"",""1.5""]";
					};
					class cargo
					{
						property="TER_VASS_cargo";
						control="VASS_AmmoBox";
						displayName="Shop Inventory";
						tooltip = "Define the items which the trader can sell and buy.\n""-"" means the item is not part of the shop, so not buyable or sellable. Script: -2.\n""∞"" means that the item will never run out. Script: -1.\n""0"" means that the item is not sold by the shop but can be sold to the trader. Script: 0.\nAny other number restricts the amonut of sellable items to the specified amount. If all items are sold out it is treated just as ""0"". Script: N.";
						expression="if (_value isEqualType '') then {\
							_value = parseSimpleArray _value;\
							_this setVariable ['TER_VASS_cargo', _value];\
							_this setVariable ['TER_VASS_cargo_default', _value];\
							};\
						";
					};
					class refresh
					{
						displayName = "Refresh";
						tooltip = "Sets the time in seconds which the shop will take to add/remove the items that were removed/added during mission. -1 disables refreshing.";
						property = "VASS_refesh";
						control = "EditShort";
						expression = "_this setVariable ['TER_VASS_refresh', _value];";
						defaultValue = "-1";
						validate = "number";
					};
					class shared
					{
						displayName = "Global";
						tooltip = "If checked the inventory's shop will broadcast its changes over the network and make it the same for everyone.";
						property = "VASS_shared";
						control = "Checkbox";
						expression = "_this setVariable ['TER_VASS_shared', _value];";
						defaultValue = "true";
					};
				};
			};
		};
	};
	class Attributes
	{
		class Default;
		class Title: Default
		{
			class Controls
			{
				class Title: ctrlStatic
				{
					style=1;
					x=0;
					w = ATTRIBUTE_TITLE_W * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
					colorBackground[]={0,0,0,0};
				};
			};
		};
		class TitleWide: Default
		{
			class Controls
			{
				class Title: ctrlStatic
				{
					x = ATTRIBUTE_CONTENT_H * GRID_W;
					w = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - ATTRIBUTE_CONTENT_H) * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
					colorBackground[]={0,0,0,0};
				};
			};
		};
		#define _YADD(VAL) __EXEC(_y = _y + VAL + 0.1)
		#define _Y __EVAL(_y)
		__EXEC(_y = 0)
		class VASS_addAction: Title
		{
			h= 5.4 * ATTRIBUTE_CONTENT_H * GRID_H;
			// Format: [Enable, Title, condition, radius, priority]
			attributeLoad="\
				_value = parseSimpleArray _value;\
				(_this controlsGroupCtrl 100) cbSetChecked (_value#0);\
				for ""_i"" from 1 to 4 do {(_this controlsGroupCtrl (100 + _i)) ctrlSetText (_value#_i);};\
			";
			attributeSave="\
				str[\
					cbChecked (_this controlsGroupCtrl 100),\
					ctrlText (_this controlsGroupCtrl 101),\
					ctrlText (_this controlsGroupCtrl 102),\
					str parseNumber ctrlText (_this controlsGroupCtrl 103),\
					str parseNumber ctrlText (_this controlsGroupCtrl 104)\
				]\
			";
			class Controls: Controls
			{
				class Title1: Title
				{
					text = "Add Shop";
					tooltip = "Enable access to the modified arsenal on this object.";
				};
				class enable: ctrlCheckboxBaseline
				{
					idc=100;
					x= ATTRIBUTE_TITLE_W * GRID_W;
					y = _Y;
					w= ATTRIBUTE_CONTENT_H * GRID_W;
					h= ATTRIBUTE_CONTENT_H * GRID_H;
				};
				_YADD(1)
				class Title2: Title
				{
					text = "addAction Title";
					tooltip = "Text shown in the scroll menu. Supports formatted text.";
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class addTitle: ctrlEdit
				{
					idc = 101;
					x = ATTRIBUTE_TITLE_W * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = ATTRIBUTE_CONTENT_W * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
				};
				_YADD(1)
				class Title3: Title
				{
					text = "Condition";
					tooltip = "Determine when the shop is accessible. Go to the addAction BIKI page for more information.";
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class addCondition: addTitle
				{
					idc = 102;
					autocomplete = "scripting";
					font="EtelkaMonospacePro";
					sizeEx="3.41 * (1 / (getResolution select 3)) * pixelGrid * 0.5";
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
				};
				_YADD(1)
				class Title4: Title
				{
					text = "Radius";
					tooltip = "Sets from how far the player can access the shop.";
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class radius: addTitle
				{
					idc = 103;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = round(ATTRIBUTE_CONTENT_W / 3) * GRID_W;
				};
				_YADD(1)
				class Title5: Title
				{
					text = "Priority";
					tooltip = "Higher values will place the scroll menu entry higher.";
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class priority: radius
				{
					idc = 104;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
				};
			};
		};
		__EXEC(_y = 0)
		class VASS_AmmoBox: TitleWide
		{
			// DEBUG
			/*
			onLoad = "[""onLoad"",_this] call compile preprocessfilelinenumbers ""gui\scripts\vasscargo.sqf"";";
			attributeLoad="[""attributeLoad"",[_this,_value,true]] call compile preprocessfilelinenumbers ""gui\scripts\vasscargo.sqf"";";
			attributeSave="[""attributeSave"",[_this]] call compile preprocessfilelinenumbers ""gui\scripts\vasscargo.sqf"";";
			*/
			// MOD
			
			onLoad = "[""onLoad"",_this] call TER_fnc_vasscargo;";
			attributeLoad = "[""attributeLoad"",[_this,_value,true]] call TER_fnc_vasscargo;";
			attributeSave = "[""attributeSave"",[_this]] call TER_fnc_vasscargo;";
			
			h = (17 * ATTRIBUTE_CONTENT_H + 1) * GRID_H;
			class Controls: Controls
			{
				class Title: Title
				{
					//text="Content";
					w = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - ATTRIBUTE_CONTENT_H - 51) * GRID_W;
				};
				class toggleView: ctrlToolbox
				{
					idc = IDC_VASSCARGO_TOOLVIEW;
					x = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - 25) * GRID_W;
					w = 25 * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
					rows = 1;
					columns = 2;
					strings[] = {"GUI", "Array"};
				};
				_YADD(1)
				class grpArrayEdit: ctrlControlsGroup
				{
					idc = IDC_VASSCARGO_GRPCARGOARRAY;
					fade = 1;
					x = ATTRIBUTE_CONTENT_H * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - ATTRIBUTE_CONTENT_H) * GRID_W;
					h = (2 + 13 - 0.1) * ATTRIBUTE_CONTENT_H * GRID_H;
					class controls
					{
						class arrayEdit: ctrlEdit
						{
							idc = IDC_VASSCARGO_EDCARGOARRAY;
							style = "0x00 + 0x10";
							font="EtelkaMonospacePro";
							sizeEx="3.41 * (1 / (getResolution select 3)) * pixelGrid * 0.5";
							x = 0;
							y = 0;
							w = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - ATTRIBUTE_CONTENT_H) * GRID_W;
							h = (2 + 13 - 0.1) * ATTRIBUTE_CONTENT_H * GRID_H;
						};
					};
				};
				/*
				class exportCargo: ctrlButton
				{
					idc = IDC_VASSCARGO_BTNEXPORT;
					text = "Export";
					tooltip = "Copy the current cargo to the clipboard.";
					x = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - 25) * GRID_W;
					w = 25 * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
				}; 
				class importCargo: exportCargo
				{
					idc = IDC_VASSCARGO_BTNIMPORT;
					text = "Import";
					tooltip = "Import a cargo array from the clipboard.";
					x = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - 51) * GRID_W;
				};
				*/
				
				class Filter: ctrlToolbox
				{
					idc=IDC_VASSCARGO_TOOLFILTER;
					style="0x30 + 0x800";
					x = ATTRIBUTE_CONTENT_H * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - ATTRIBUTE_CONTENT_H) * GRID_W;
					h = 2 * ATTRIBUTE_CONTENT_H * GRID_H;
					rows=1;
					columns=13;
					strings[]=
					{
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_0_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_1_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_2_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_3_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_4_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_5_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_6_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_7_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_8_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_9_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_10_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_11_ca.paa",
						"\a3\Ui_F_Curator\Data\RscCommon\RscAttributeInventory\filter_12_ca.paa"
					};
				}; _YADD(1.9)
				class ListBackground: Title
				{
					idc = IDC_VASSCARGO_STATICLISTBACKGROUND;
					x = ATTRIBUTE_CONTENT_H * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - ATTRIBUTE_CONTENT_H) * GRID_W;
					h = 13 * ATTRIBUTE_CONTENT_H * GRID_H;
					colorBackground[]={1,1,1,0.1};
				};
				class List: ctrlListNBox
				{
					idc=IDC_VASSCARGO_LNBCARGO;
					//style = 32;
					x = ATTRIBUTE_CONTENT_H * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - ATTRIBUTE_CONTENT_H) * GRID_W;
					h = 13 * ATTRIBUTE_CONTENT_H * GRID_H;
					drawSideArrows=1;
					idcLeft=IDC_VASSCARGO_BTNMINUS;
					idcRight=IDC_VASSCARGO_BTNPLUS;
					columns[]={0.050000001,0.15000001,0.7,0.85000002};
					disableOverflow=1;
				}; _YADD(12.9)
				class Validate: ctrlStructuredText
				{
					idc = IDC_VASSCARGO_TXTVALIDATE;
					x = ATTRIBUTE_CONTENT_H * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - ATTRIBUTE_CONTENT_H) * GRID_W;
					h = 1 * ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class BtnValidate: ctrlButton
				{
					idc = IDC_VASSCARGO_BTNVALIDATE;
					text = "Validate";
					x = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - 25) * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = 25 * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class TextPrice: Title
				{
					idc = IDC_VASSCARGO_TITLEPRICE;
					text = "Price:";
					style=0;
					x = ATTRIBUTE_CONTENT_H * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = 15 * GRID_W;
				};
				class Price: ctrlEdit
				{
					idc=IDC_VASSCARGO_EDPRICE;
					x = 20 * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = 25 * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class ButtonClear: ctrlButton
				{
					idc=IDC_VASSCARGO_BTNCLEAR;
					text="Clear";
					x = (ATTRIBUTE_TITLE_W + ATTRIBUTE_CONTENT_W - 25) * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = 25 * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
				};/* _YADD(1)
				
				class TextFactionFilter: Title
				{
					text = "Filter by faction";
					tooltip = "Only show items which the given faction uses.";
					style = 1;
					x = 0;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = ATTRIBUTE_TITLE_W * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class FactionFilter: ctrlCombo
				{
					idc = IDC_VASSCARGO_COMBOPRESETS;
					x = ATTRIBUTE_TITLE_W * GRID_W;
					y = _Y * ATTRIBUTE_CONTENT_H * GRID_H;
					w = ATTRIBUTE_CONTENT_W * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
				};
				*/
				class ArrowLeft: ctrlButton
				{
					idc=IDC_VASSCARGO_BTNMINUS;
					text="-";
					font="RobotoCondensedBold";
					x = -1;
					y = -1;
					w = ATTRIBUTE_CONTENT_H * GRID_W;
					h = ATTRIBUTE_CONTENT_H * GRID_H;
				};
				class ArrowRight: ArrowLeft
				{
					idc=IDC_VASSCARGO_BTNPLUS;
					text="+";
				};
			};
		};
	};
};